universal-login-commons/src/integration/ethereum/ProviderService.ts

```diff
@@ -4,11 +4,19 @@ import {fetchHardforkVersion} from './fetchHardforkVersion';
 import {NetworkVersion} from '../../core/utils/messages/computeGasData';
 
 export class ProviderService {
+  private cachedContractCodes: Record<string, Promise<string>> = {};
+
   constructor(private provider: providers.Provider) {
   }
 
   getCode(contractAddress: string) {
-    return this.provider.getCode(contractAddress);
+    return this.cachedContractCodes[contractAddress] || this.getCodeFromProviderAndCacheIt(contractAddress);
+  }
+
+  private getCodeFromProviderAndCacheIt(contractAddress: string) {
+    const code = this.provider.getCode(contractAddress);
```

Comment:
getCode is async, isn't it?

Reply:
Yeah, but we are storing and returning Promise<string>

Reply:
Okej, done

Reply:
Use await here